File: PxGraf.Frontend/src/types/editorContentsResponse.ts
import { MultiLanguageString } from "./multiLanguageString";
import { VisualizationType } from "./visualizationType";

/**
 * Interface for dimension sorting options.
 * @property {string} code - The code of the sorting option.
 * @property {MultiLanguageString} description - Longer, user facing description of the sorting option.
 */
export interface ISortingOption {
    code: string;
    description: MultiLanguageString;
}

/**
 * Interface for sorting option collection. Contains default and pivoted sorting options if available.
 * @property {ISortingOption[]} default - Default sorting options.
 * @property {ISortingOption[]} pivoted - Pivoted sorting options.
 */
interface ISortingOptionCollection {
    default: ISortingOption[];
    pivoted: ISortingOption[];
}

/**
 * Interface for visualization options for a specific visualization type.
 * @property {VisualizationType} type - The type of visualization.
 * @property {boolean} [allowShowingDataPoints] - Flag to indicate if showing data points is allowed.
 * @property {boolean} [allowCuttingYAxis] - Flag to indicate if cutting the Y-axis is allowed.
 * @property {boolean} [allowMatchXLabelsToEnd] - Flag to indicate if matching X labels to the end is allowed.
 * @property {boolean} [allowSetMarkerScale] - Flag to indicate if setting marker scale is allowed.
 * @property {boolean} allowManualPivot - Flag to indicate if manual pivoting is allowed.
 * @property {boolean} allowMultiselect - Flag to indicate if multivalue selectable is allowed.
 * @property {ISortingOptionCollection} sortingOptions - Sorting options for the visualization containing default and pivoted options.
 */
export interface IVisualizationOptions {
    type: VisualizationType;
    allowShowingDataPoints?: boolean;
    allowCuttingYAxis?: boolean;
    allowMatchXLabelsToEnd?: boolean;
    allowSetMarkerScale?: boolean;
    allowManualPivot: boolean;
    allowMultiselect: boolean;
    sortingOptions: ISortingOptionCollection;
}

/**
 * Interface for the response from the editor contents API used to populate the settings for adjusting the visualization in the editor.
 * @property {number} size - The size of the response based on the amount of values selected from all dimensions.
 * @property {number} sizeWarningLimit - The size warning limit for the response.
 * @property {number} maximumSupportedSize - The maximum supported size for the response.
 * @property {MultiLanguageString} headerText - Default header text for the visualization as a @ {@link MultiLanguageString} object.
 * @property {number} maximumHeaderLength - The maximum length of the header text.
 * @property {IVisualizationOptions[]} visualizationOptions - Visualization options for the valid visualization types as an array of @ {@link IVisualizationOptions} objects.
 * @property {object} visualizationRejectionReasons - Object containing rejection reasons for each visualization type as a @ {@link MultiLanguageString} object.
 */
export interface IEditorContentsResponse {
    size: number;
    sizeWarningLimit: number;
    maximumSupportedSize: number;
    headerText: MultiLanguageString;
    maximumHeaderLength: number;
    visualizationOptions: IVisualizationOptions[];
    visualizationRejectionReasons: {
        [visualizationType: string]: MultiLanguageString;
    }
}

File: PxGraf/Models/Responses/EditorContentsResponse.cs
#nullable enable
using Px.Utils.Language;
using PxGraf.Enums;
using System.Collections.Generic;

namespace PxGraf.Models.Responses
{
    public class EditorContentsResponse
    {
        /// <summary>
        /// Current size of the query.
        /// </summary>
        public required int Size { get; set; }

        /// <summary>
        /// The limit when the end user should be warned that the size of the query is getting larger than is practical.
        /// </summary>
        public required int SizeWarningLimit { get; set; }

        /// <summary>
        /// The limit when the end user should be told that the size of the query can cause technical issues and further actions will be prevented unless the size is reduced
        /// </summary>
        public required int MaximumSupportedSize { get; set; }

        /// <summary>
        /// Default header for the query.
        /// </summary>
        public required MultilanguageString HeaderText { get; set; }

        /// <summary>
        /// Maximum length of the header text, in any language
        /// </summary>
        public required int MaximumHeaderLength { get; set; }

        /// <summary>
        /// These visualization types are valid for the query
        /// </summary>
        public required List<VisualizationOption> VisualizationOptions { get; set; }

        /// <summary>
        /// Collection of reasons why the invalid visualization types are not possible for this query.
        /// </summary>
        public required Dictionary<VisualizationType, MultilanguageString> VisualizationRejectionReasons { get; set; }

        /// <summary>
        /// Indicates whether publication webhooks are enabled in the configuration.
        /// When false, queries should be saved as published (non-draft) by default since there's no webhook to trigger.
        /// </summary>
        public required bool PublicationWebhookEnabled { get; set; }
    }

    public class VisualizationOption
    {
        public required VisualizationType Type { get; set; }

        public class SortingOptionsCollection
        {
            public required List<SortingOption>? Default { get; set; }
            public required List<SortingOption>? Pivoted { get; set; }
        }

        public required bool AllowShowingDataPoints { get; set; }
        public required bool AllowCuttingYAxis { get; set; } 
        public required bool AllowMatchXLabelsToEnd { get; set; }
        public required bool AllowSetMarkerScale { get; set; }

        /// <summary>
        /// Whether manual ordering of dimensions is allowed.
        /// </summary>
        public required bool AllowManualPivot { get; set; }

        /// <summary>
        /// List of available sorting options.
        /// </summary>
        public required SortingOptionsCollection SortingOptions { get; set; }

        /// <summary>
        /// Whether the user is allowed to define dimensions with multiple selectable values displayed simultaneously.
        /// </summary>
        public required bool allowMultiselect { get; set; }
    }
}
#nullable disable
